Refactor: unify stellarator configuration loaders into a single get_data API

armulrich opened the pull request:
This change consolidates all of the individual get_*_data loaders into a single get_data function, streamlining the API and mirroring the style used in the [pyQSC repository](https://github.com/landreman/pyQSC/blob/cd75359ea47548d5db7ccb458c100085c04ba1bc/qsc/configurations.py#L9). It updates the test suite to call get_data instead of the old functions, preserves backward compatibility via thin wrappers, and refreshes the module’s documentation so that all configurations (“ncsx,” “hsx,” “giuliani,” and “w7x”) are described consistently in one place. It was raised by @landreman  and @mishapadidar.

landreman commented on tests/geo/test_surface_objectives.py:
```diff
@@ -64,7 +64,7 @@ def test_toroidal_flux_is_constant(self):
         """
         s = get_exact_surface()
-        curves, currents, ma = get_ncsx_data()
+        curves, currents, ma, _ = get_data("ncsx")
         nfp = 3
         coils = coils_via_symmetries(curves, currents, nfp, True)
         bs_tf = BiotSavart(coils)
```
Lines 68-70 could be eliminated now because the BiotSavart is returned directly by get_data(). This is also true of many of the other places that get_data is called.

armulrich replied:
I agree. Will address this.

landreman commented on tests/geo/test_qfm.py:
```diff
@@ -49,13 +49,14 @@ def test_qfm_objective_gradient(self):
+    def subtest_qfm_objective_gradient(self, surfacetype, stellsym, config):
         np.random.seed(1)
-        curves, currents, ma = get_data()
+        curves, currents, ma, _ = get_data(config)
         nfp = ma.nfp
         coils = coils_via_symmetries(curves, currents, nfp, True)
         bs = BiotSavart(coils)
```
I believe lines 60-62 can be eliminated now since get_data returns a BiotSavart object. I believe line 63 could be replaced with `bs_tf = BiotSavart(bs.coils)` or perhaps some kind of copy() could be used.

armulrich replied:
Yep, agree with this too.

armulrich replied:
Finished implementing this too.

landreman reviewed (commented):
Looks great! Thanks for working on this.

Can we also add `simsopt.configs.LHD_like` to the same get_config() function?

It might be nice to return nfp for each config as well. (I see in the tests that a hack for this is to get the nfp attribute of the magnetic axis, but this trick isn't obvious). Would it make sense to include nfp as another returned quantity?

armulrich commented:
> Looks great! Thanks for working on this.
> 
> Can we also add `simsopt.configs.LHD_like` to the same get_config() function?
> 
> It might be nice to return nfp for each config as well. (I see in the tests that a hack for this is to get the nfp attribute of the magnetic axis, but this trick isn't obvious). Would it make sense to include nfp as another returned quantity?

No problem @landreman ! Glad to hear it's going in the right direction.

And yes, I think adding `simsopt.configs.LHD_like` makes a lot of sense and helps clean up the configuration interface further. I'll suggest removing the standalone `get_LHD_like_data` function and instead integrating it directly into `zoo.py`, registering it as `"lhd_like"`. Also, just to note: I called the general function `get_data` rather than `get_config`. To me, it felt a bit clearer and flows better syntactically (eg. `get_data("ncsx")`). Let me know if you think there's a strong reason to prefer `get_config` instead.

As for returning `nfp`, I completely agree, it avoids the need to unwrap it from `ma.nfp`, which isn’t always obvious to users. I’d propose returning it explicitly like this from `get_data`:

```python
return curves, currents, ma, nfp, bs
```

Note: I placed `nfp` before `bs` since it conceptually follows from the magnetic axis definition, hope you agree with that.

armulrich commented:
Addressed all of the above comments in this commit.

smiet commented:
Some people have scripts that use on these functions. 

- Would it be an idea to keep the old functions (until the next major release) and add a deprecationwarning, similar to what is done in NumPy?

- We are also working on a QUASR loader, that automatically downloads any QUASR id. This has a different return style (QUASR does not save the axis and so cannot return it) Does it still make sense to consolidate this into a unified configuration-getting funciton? @andrewgiuliani @landreman

mishapadidar commented on src/simsopt/configs/zoo.py:
```diff
@@ -2,188 +2,334 @@
 
-
-def get_ncsx_data(Nt_coils=25, Nt_ma=10, ppp=10):
+def get_data(name, **kwargs):
```
what do we think about the name `get_data`? should we rename to something like `load_config`?

Similarly, we could change `name` to something more descriptive like `configuration` or `design`.

armulrich replied:
I’m fine with get_data, it’s short and to the point, but ``load_config`` would perhaps make the intent a bit clearer. Happy to switch if that feels more intuitive to everyone. @landreman perhaps you have a take on this.

armulrich replied:
@mishapadidar No further concerns were raised, so I’ve kept it as `get_data` for now.

mishapadidar commented on src/simsopt/configs/zoo.py:
```diff
@@ -2,188 +2,334 @@
+    name : str
+        Which configuration to load. Available values are:
+        ``"ncsx"``, ``"hsx"``, ``"giuliani"``, ``"w7x"``, ``"lhd_like"``.
+    Nt_coils : int, optional
```
eventually we should consider renaming this arguments. `Nt_coils` and `ppp` are not great names.

mishapadidar replied:
we should change the naming now so we dont get locked in to a bad convention.

armulrich replied:
Addressed this in the latest push.

mishapadidar commented on src/simsopt/configs/zoo.py:
```diff
@@ -2,188 +2,334 @@
+        (*LHD_like only*) Number of quadrature points for the six circular coils.
+    numquadpoints_helical : int, optional
+        (*LHD_like only*) Number of quadrature points for each helical coil.
+    numquadpoints_axis : int, optional
```
I could see this list of arguments becoming quite long as we add more configurations with their own set of parameters.

mishapadidar replied:
The kwargs handles this in a clean way in the code. The only problem is the documentation getting messy. One way to handle this is by setting up the documentation to be like the scipy minimize. If this is not possible, then another way is to just add sections in the documentation devoted to each config. Something like, 
```
Parameters
    ----------
    name : str
        Which configuration to load. Available values are:
        ``"ncsx"``, ``"hsx"``, ``"giuliani"``, ``"w7x"``, 
    kwargs: dict
        Configuration specific parameters. See documentation of these parameters below.


ncsx
    ----------
    blah : discussion of parameters.
```

This might be overkill for now though, since we only have a small number of configurations.

armulrich replied:
Yeah, kwargs keeps the code tidy. The mess is purely in the docs. I’m fine slimming the main “Parameters” block to kwargs: dict and then listing config‑specific options in short subsections (“ncsx”, “hsx”, …) right below, like you suggest. If we later spin up separate pages a la scipy.optimize.minimize, we can move those details there and just link out. For now I’ll probaly do the lightweight subsection approach then...

armulrich replied:
Resolved by slimming the main `Parameters` block to `kwargs: dict` and moving config-specific arguments into short subsections for each configuration.

mishapadidar commented on src/simsopt/configs/zoo.py:
```diff
@@ -2,188 +2,334 @@
+    Returns
+    -------
+    tuple
+        *5-element tuple* ``(curves, currents, ma, nfp, bs)``, where:
```
curves and currents should already be in the `bs` object right? perhaps we can remove them from the tuple. `nfp` should also be in the `bs` or `ma` right?

mishapadidar replied:
hmm, okay. maybe we leave `nfp` then? I am still in favor of removing curves and currents. Lets get more opinions @andrewgiuliani @landreman

armulrich replied:
Leaving tuple contents unchanged for now to preserve the old API and maintain convenience access to `curves` and `currents`. `nfp` is still included since it is not exposed directly by `bs` or `ma` (and definitions differ for `lhd_like`). No changes made in the last push. Can revisit if consensus shifts toward removing `curves`/`currents`. @mishapadidar @landreman @andrewgiuliani

smiet replied:
I am happy with the current return tuple, but I think the description and naming should be changed to `base_curves`, `base_currents` to emphasize that this returns only the unique coils. Providing these is useful because it is not straighforward to extract these from the bs, which contains ss and nfp copies of them.

mishapadidar commented on src/simsopt/configs/zoo.py:
```diff
@@ -2,188 +2,334 @@
+            
+                - For all configurations **except** ``"lhd_like"``, the coils are expanded via ``coils_via_symmetries(curves, currents, nfp, True)`` before building ``bs``.
+                
+                - For ``"lhd_like"``, each ``CurveXYZFourier`` and its ``Current`` are passed directly into ``Coil(curve, current)`` (no symmetry expansion).
```
Perhaps it is worth separating the documentation a bit, so that any notes specific to configuration arguments or output are listed alongside the description of the configuration. what do you think?

mishapadidar replied:
Question: for `lhd_like` and all other configurations, do we always return the full set of coils? LHD is the only configuration with helical coils so perhaps the coils already represent the full coil set. If that is so, then we can remove this comment on the case-by-case behavior and just say `we return a biot savart for all coils`. If this is not the case, then maybe we should modify the behavior of `lhd_like` to return the full set of coils.

mishapadidar commented on src/simsopt/configs/zoo.py:
```diff
@@ -2,188 +2,334 @@
-
-def get_giuliani_data(Nt_coils=16, Nt_ma=10, ppp=10, length=18, nsurfaces=5):
-    """
+    Notes
```
I think a good example to follow in the documentation is the [scipy minimize](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html). For scipy minimize, the `method` argument has many options, similar how we have many options for `name`. You'll notice that for each method, there is a link to a separate documentation page discussing the specific arguments and quirks of using scipy minimize with that method. For example, the Nelder Mead optimizer has its own [page](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-neldermead.html#optimize-minimize-neldermead). 

We could try to reproduce this here. For each configuration we have a link to a separate documentation page that discusses the specific arguments, and the modifications to the output tuple.

armulrich replied:
That sounds like a solid approach. We can keep the high-level overview and shared parameters in the main get_data docstring, then link out to individual pages, one per config, where we dive into the nitty gritty params and return.

mishapadidar replied:
If this is too difficult we can drop it.

armulrich replied:
Ok, got it.

armulrich commented:
@smiet Addressed all of the following in the latest push.  
> Looks like there are a few more changes to do before we can move forward:
> 
> * [x]  re-instate the old `get_<name>_data` scripts with a `DeprecationWarning`
> * [x]  change the naming of `ppp` and `Nt_X` (as per discussion^: [Refactor: unify stellarator configuration loaders into a single get_data API #532 (comment)](https://github.com/hiddenSymmetries/simsopt/pull/532#discussion_r2222969306))
> * [x]  Resolve the merge conflict with the updated master.
> 
> [When this PR is merged, I will adapt #425 to use this function.]


Extra notes:
- Restored old `get_*_data` functions as well as `get_LHD_like`, with deprecation warnings added to the docstrings for clarity.
- Updated `get_data` parameter naming conventions to improve readability.
<img width="718" height="783" alt="Screenshot 2025-08-09 at 14 28 04" src="https://github.com/user-attachments/assets/6cee68a9-0de2-47a9-9340-f8c7791fd68a" />
<img width="472" height="825" alt="Screenshot 2025-08-09 at 14 30 33" src="https://github.com/user-attachments/assets/c93d2c1e-1f7b-4bfb-aadd-bac1b2d8ed30" />

<img width="492" height="425" alt="Screenshot 2025-08-09 at 14 30 43" src="https://github.com/user-attachments/assets/eea52ba8-0be3-4902-bfbf-46c3e9555903" />

<img width="347" height="721" alt="Screenshot 2025-08-09 at 14 28 47" src="https://github.com/user-attachments/assets/6ecb16d5-0442-4b00-bf61-ae7fa580bd0d" />

<img width="342" height="796" alt="Screenshot 2025-08-09 at 14 28 57" src="https://github.com/user-attachments/assets/da39d009-91c5-45d5-9f95-470c257f3476" />

smiet commented on tests/configs/test_LHD_like.py:
```diff
@@ -12,12 +12,11 @@ def test_axis(self):
-            Coil(curve, Current(current)) for curve, current in zip(coils, currents)
+            Coil(curve, -1*current) for curve, current in zip(curves, currents)
         ]
         field = BiotSavart(coils)
```
use `bs` returned from get_data.

armulrich replied:
My concern with this outlined below (minus phi is a special case)

smiet commented on tests/field/test_coilset.py:
```diff
@@ -95,10 +93,9 @@ def test_base_coils(self):
-        new_base_coils = [Coil(curve, current) for curve, current in zip(curves, currents)]
-        self.coilset.base_coils = new_base_coils
-        self.assertEqual(self.coilset.base_coils, new_base_coils)
+        curves, currents, ma, nfp, bs = get_data("ncsx")
```
A CoilSet should be provided with base_coils (the unique coils). This passes but uses the setter incorrectly. 

replace with
```
        curves, currents, ma, nfp, bs = get_data("ncsx")
        new_base_coils = [Coil(curve, current) for curve, current in zip(curves, currents)]
        self.coilset.base_coils = new_base_coils
        self.assertEqual(self.coilset.base_coils, new_base_coils)
        ```

armulrich replied:
implemented this too.

smiet commented on tests/geo/surface_test_helpers.py:
```diff
@@ -84,10 +84,9 @@ def get_boozer_surface(label="Volume", nphi=None, ntheta=None, boozer_type='exac
-    current_sum = sum(abs(c.current.get_value()) for c in coils)
+    base_curves, base_currents, ma, nfp, bs = get_data("ncsx")
+    
+    current_sum = sum(abs(c.current.get_value()) for c in bs.coils)
```
use returned currents * nfp instead of all coils in bs

smiet commented on tests/geo/surface_test_helpers.py:
```diff
@@ -116,7 +114,7 @@ def get_boozer_surface(label="Volume", nphi=None, ntheta=None, boozer_type='exac
         lab_target = lab.J()
     elif label == "ToroidalFlux":
-        bs_tf = BiotSavart(coils)
+        bs_tf = BiotSavart(bs.coils)
```
just use `bs` object?

armulrich replied:
same issue as pointed out already

smiet commented on tests/geo/test_boozersurface.py:
```diff
@@ -24,10 +24,8 @@ def test_residual(self):
-        bs = BiotSavart(coils)
-        bs_tf = BiotSavart(coils)
+        curves, currents, ma, nfp, bs = get_data("ncsx")
+        bs_tf = BiotSavart(bs.coils)
```
use bs object

smiet commented on tests/geo/test_boozersurface.py:
```diff
@@ -82,11 +80,9 @@ def test_boozer_penalty_constraints_hessian(self):
-        current_sum = sum(abs(c.current.get_value()) for c in coils)
+        curves, currents, ma, nfp, bs = get_data("ncsx")
+        bs_tf = BiotSavart(bs.coils)
+        current_sum = sum(abs(c.current.get_value()) for c in bs.coils)
```
only unique currents

armulrich replied:
done

smiet commented on tests/geo/test_boozersurface.py:
```diff
@@ -395,12 +387,10 @@ def subtest_convergence_cpp_and_notcpp_same(self, vectorize):
-        current_sum = sum(abs(c.current.get_value()) for c in coils)
-        bs = BiotSavart(coils)
+        curves, currents, ma, nfp, bs = get_data("ncsx")
+        current_sum = sum(abs(c.current.get_value()) for c in bs.coils)
```
only unique currents

armulrich replied:
done

smiet commented on tests/geo/test_boozersurface.py:
```diff
@@ -447,11 +437,9 @@ def test_boozer_penalty_constraints_cpp_notcpp(self):
-        bs_tf = BiotSavart(coils)
-        current_sum = sum(abs(c.current.get_value()) for c in coils)
+        curves, currents, ma, nfp, bs = get_data("ncsx")
+        bs_tf = BiotSavart(bs.coils)
```
use returned bs and only unique currents

smiet commented on tests/geo/test_qfm.py:
```diff
@@ -25,11 +25,8 @@ def test_residual(self):
-        bs = BiotSavart(coils)
-        bs_tf = BiotSavart(coils)
+        curves, currents, ma, nfp, bs = get_data("ncsx")
+        bs_tf = BiotSavart(bs.coils)
```
same bs

armulrich replied:
same as above

smiet commented on tests/geo/test_qfm.py:
```diff
@@ -49,17 +46,15 @@ def test_qfm_objective_gradient(self):
-        bs = BiotSavart(coils)
-        bs_tf = BiotSavart(coils)
+        curves, currents, ma, nfp, bs = get_data(config)
+        bs_tf = BiotSavart(bs.coils)
```
same bs

smiet commented on tests/geo/test_qfm.py:
```diff
@@ -100,11 +95,8 @@ def test_qfm_label_constraint_gradient(self):
-        bs = BiotSavart(coils)
-        bs_tf = BiotSavart(coils)
+        curves, currents, ma, nfp, bs = get_data("ncsx")
+        bs_tf = BiotSavart(bs.coils)
```
same bs

smiet commented on tests/geo/test_qfm.py:
```diff
@@ -140,17 +132,15 @@ def test_qfm_penalty_constraints_gradient(self):
-        bs = BiotSavart(coils)
-        bs_tf = BiotSavart(coils)
+        curves, currents, ma, nfp, bs = get_data(config)
+        bs_tf = BiotSavart(bs.coils)
```
same

armulrich replied:
`tf` issue outlined below

smiet commented on tests/geo/test_surface_objectives.py:
```diff
@@ -64,10 +63,8 @@ def test_toroidal_flux_is_constant(self):
-        coils = coils_via_symmetries(curves, currents, nfp, True)
-        bs_tf = BiotSavart(coils)
+        curves, currents, ma, nfp, bs= get_data("ncsx")
+        bs_tf = BiotSavart(bs.coils)
```
same bs

smiet reviewed (changes requested):
Thank you for the work, it is looking really good! sorry for the delay in my review (vacation). 

There a few more things: 
-The `curves` and `currents` returned by `get_data` are only the unique coils, whereas `bs.coils` contains the symmetry-copies. can you make this clear by calling these `base_curves` and `base_currents` in documentation and in examples (if you are a regex ninja also in tests)
- you often create a new BiotSavart object with `bs_tf= BiotSavart(bs.coils)` these two should be identical, and it is simpler and clearer to use the `bs` that is returned by `get_data`
- The `current_sum` object that you create by summing all currents in `bs.coils` will sum the same value many times. Best practice is to only take the unique values and sum them (multiplied by 2*nfp if the absolute value is important).

armulrich commented on examples/1_Simple/qfm.py:
```diff
@@ -17,15 +17,12 @@
-bs = BiotSavart(coils)
-bs_tf = BiotSavart(coils)
+curves, currents, ma, nfp, bs = get_data("ncsx")
+bs_tf = BiotSavart(bs.coils)
```
NOTE: this impacts all the subsequent use same ``bs_tf = bs`` comments

armulrich commented:
Thanks again for the thorough feedback @smiet.  
I updated the code to consistently use `base_curves`, `base_currents`, and `all_curves`, and fixed all instances of `current_sum` so that only unique currents are counted.  

The two items I haven’t implemented (due to concerns) are:  
1. **LHD-like config:** I didn’t switch to using the returned `bs` object, since `get_data` currently builds `bs` with positive currents. Flipping the sign afterward gives a different `bs` than the one used in the test, which causes it to fail. Hence need to build separate ``bs`` object.
2. **`bs_tf` vs single `bs`:** Throughout the `get_data` usage kept the second `BiotSavart` object (`bs_tf`). A single `bs` isn’t reliable because `BiotSavart.set_points()` is mutable. `QfmResidual` and `ToroidalFlux` require different grids at the same time, and reusing one `bs` causes reshape errors. Having a duplicate `BiotSavart` avoids this conflict while still referencing the same coil set.  

If you’d like me to handle either of these two points differently, let me know.

armulrich commented on examples/2_Intermediate/boozer.py:
```diff
@@ -19,18 +19,15 @@
-bs_tf = BiotSavart(coils)
-current_sum = sum(abs(c.current.get_value()) for c in coils)
+curves, currents, ma, nfp, bs = get_data("ncsx")
+curves = [c.curve for c in bs.coils]
```
implemented this below with comment

armulrich commented on examples/2_Intermediate/boozer.py:
```diff
@@ -19,18 +19,15 @@
+curves, currents, ma, nfp, bs = get_data("ncsx")
+curves = [c.curve for c in bs.coils]
+bs_tf = BiotSavart(bs.coils)
+current_sum = sum(abs(c.current.get_value()) for c in bs.coils)
```
fixed below

armulrich commented on tests/geo/test_boozersurface.py:
```diff
@@ -122,11 +118,9 @@ def subtest_boozer_penalty_constraints_gradient(self, surfacetype, stellsym,
-        bs_tf = BiotSavart(coils)
-        current_sum = sum(abs(c.current.get_value()) for c in coils)
+        curves, currents, ma, nfp, bs = get_data("ncsx")
+        bs_tf = BiotSavart(bs.coils)
```
implemented partly with current sum fixed but kept `bs_tf` the same due to issue outlined below
